sql/analyzer/apply_hash_lookups.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+)
+
+func applyHashLookups(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+	return plan.TransformUpWithParent(n, func(n sql.Node, parent sql.Node, childNum int) (sql.Node, error) {
+		cr, isCachedResults := n.(*plan.CachedResults)
+		pj, _ := parent.(plan.JoinNode)
+		pij, _ := parent.(*plan.IndexedJoin)
+		var cond sql.Expression
+		var primaryGetter, secondaryGetter func (sql.Expression) sql.Expression
+		if isCachedResults && pij != nil && childNum == 1 {
+			primary := pij.Left()
+			cond = pij.Cond
+			primaryIndex := len(primary.Schema()) + len(scope.Schema())
+			primaryGetter = getFieldIndexRange(0, primaryIndex, 0)
+			secondaryGetter = getFieldIndexRange(primaryIndex, -1, primaryIndex)
+		} else if isCachedResults && pj != nil && pj.JoinType() != plan.JoinTypeRight && childNum == 1 {
+			primary := pj.Left()
+			cond = pj.JoinCond()
+			primaryIndex := len(primary.Schema()) + len(scope.Schema())
+			primaryGetter = getFieldIndexRange(0, primaryIndex, 0)
+			secondaryGetter = getFieldIndexRange(primaryIndex, -1, primaryIndex)
+		} else if isCachedResults && pj != nil && pj.JoinType() == plan.JoinTypeRight && childNum == 0 {
+			// The columns from the primary row are on the right.
```

Comment:
On first reading I though this was wrong, and you could have an IndexedJoin node here, because I thought that IndexedJoin implemented JoinNode.

This logic would be simpler to read with a case switch, I think